### ISSUE TITLE
vcpkg should set VCPKG_OSX_DEPLOYMENT_TARGET from CMAKE_OSX_DEPLOYMENT_TARGET in manifest mode

### ISSUE THREAD

**Issue by MarkCallow** [microsoft/vcpkg#39981]
**Describe the bug**
On Apple OS's, If one's iOS app's deployment target is older than the deployment target of the dependencies built by vcpkg `ld` puts out hundreds of warnings. In manifest mode where vcpkg is automatically selecting the triplet, it selects a triplet without `VCPKG_OSX_DEPLOYMENT_TARGET` and builds the dependencies with the latest version of iOS as the deployment target. If your app has an earlier deployment target - not unusual - you will be subjected to the aforesaid hundreds of warnings.

**Environment**
- OS: macOS
- Compiler: Apple clang version 15.0.0 (clang-1500.3.9.4)

**To Reproduce**
1. Set up manifest mode to provide dependencies for your iOS app that uses CMake for building.
2. Set CMAKE_OSX_DEPLOYMENT_TARGET to something other than the latest version of iOS.
3. Configure and generate a build for Xcode or other tool that uses Apple clang.
4. Build your app.
5. Observe the warnings (332 in my case).

**Expected behavior**
vcpkg should use the value of `CMAKE_OSX_DEPLOYMENT_TARGET` when setting up the triplet for the build.

**Failure logs**
Here is a sample warning message:
```
Object file (<project dir>/build/ios/vcpkg_installed/arm64-ios/lib/libassimp.a[4](Compression.cpp.o)) was built for newer 'iOS' version (17.5) than being linked (12.0)
```

**Additional context**
I am aware thanks to https://github.com/microsoft/vcpkg/discussions/39966 that this can be worked around using VCPKG_OVERLAY_TRIPLETS. To do so. I have to copy a minimum of 4 triplet files for iOS and OSX, edit each of them to add VCPKG_OSX_DEPLOYMENT_TARGET to a target matching the CMAKE_OSX_DEPLOYMENT_TARGET in my CMakeLists.txt. If I need to change the deployment target in future I have to change it in multiple files with the concomitant risk of error all the while potentially missing updates to the source triplet files. I find it extremely user unfriendly to have to go to such lengths to avoid a barrage of warnings.

**Comment by dg0yt** [comment]
FTR this aren't just build-time warnings. The libs from the ports are likely to make the final app fail on the desired deployment target.

**Comment by MarkCallow** [comment]
What's the question?

**Comment by MarkCallow** [comment]
This is waiting for somebody to implement it. There is nothing further to discuss. I don't appreciate having to keep posting comments to keep an issue alive.

I would still like to know why this has been labelled "category:question" @FrankXie05.

**Comment by davidbien** [comment]
This is a problem. I need it fixed as well. Here's another example:
Object file (/Users/...Xbuild/vcpkg_installed/arm64-ios/debug/lib/libnlohmann_json_schema_validator.a[5](json-validator.cpp.o)) was built for newer 'iOS' version (18.1) than being linked (15.6)
I have 970 such warnings. I doubt it is the reason I can't get my app to build on App Store Connect but as much effort as I have put into that it would be nice to rid this warnings to rule them out. It's messy, and the only way to not have them show up is to send a blanket "-w" to ld which supresses all warnings./

**Comment by howprice** [comment]
I am also experiencing this problem. In my situation we are using vcpkg and cmake as part of the build process to generate a Universal Binary containing both x86 and arm (Apple Silicon) binaries. We are explicitly setting CMAKE_OSX_DEPLOYMENT_TARGET to maximise compatibility, but this is causing the warnings mentioned in the original post which I am not comfortable shipping with.

Thanks in advance for any efforts to resolve this issue.

**Comment by MarkCallow** [comment]
To @davidbien and @howprice these are not just warnings. See https://github.com/microsoft/vcpkg/issues/39981#issuecomment-2238522224. It most likely IS the reason you can't get your app to build on App Store Connect, @davidbien.

**Comment by davidbien** [comment]
@MarkCallow Nah, the warnings are innocuous, at least in my case, as the execution of my application doesn't depend on the actual version of iOS. If there were a true incompatibility then the app wouldn't link. Apple doesn't care if your app spews warnings on compilation. My problem was something else entirely. Of more concern, however is the spamming of these innocuous warnings will hide warnings that one should be addressing. I'm gonna do my due diligence and ensure, before any release, that I filter any innocuous warnings and address the important ones.
It's a hard problem to solve, so I get the resistance, also most of the time, as I said, if the app is truly dependent on a version of iOS then it will fail to compile or link - we are talking about embedded C++ libraries here.

**Comment by davidbien** [comment]
> FTR this aren't just build-time warnings. The libs from the ports are likely to make the final app fail on the desired deployment target.

This entirely not true. They are just build-time warnings. I've deployed many many times in the App Store Connect. Apple doesn't compile your app, you compile your app. You give your compiled app to Apple and Apple then approves the app or not, but Apple never compiles the app. I get that the warnings are annoying, but playing it up to the point of "the sky is falling" isn't helpful. The warnings are completely innocuous, if they weren't then the app would fail to link - and then there would be a problem.

**Comment by dg0yt** [comment]
If you strongly believe that the warnings are completely innocuous, please tell Apple to stop printing them.

**Comment by SanderVocke** [comment]
I think the discussion about warnings is a bit beside the point. This setting is about configuring the versions of MacOS that the compiled binaries will be guaranteed to run on. Whether some libraries still run regardless of the setting is also beside the point.

I agree with the original poster: the only feasible way to set this variable seems to be in the triplet file, so for any "regular" build of a package in manifest mode with backwards compatibility to older MacOS versions, we have to work with additional custom triplet files, OVERLAY_TRIPLETS etc. It is very cumbersome and makes CI scripts complicated.

This even causes "default" builds to result in binaries not compatible for the native host system, if there is a newer XCode version installed than the MacOS version you are building on.

I don't know if adopting CMAKE_OSX_DEPLOYMENT_TARGET is the solution though. How about respecting MACOSX_DEPLOYMENT_TARGET env var?
